tools/audio/soundfont_compiler.c

```diff
@@ -0,0 +1,1811 @@
+    sf.samples = NULL;
+    LL_FOREACH(xmlNodePtr, node, root->children) {
+        const char *name = XMLSTR_TO_STR(node->name);
+
+        if (strequ(name, "Samples"))
+            read_samples_info(&sf, node);
+    }
+
+    // read all instruments
+    memset(sf.program_number_bitset, 0, sizeof(sf.program_number_bitset));
+    sf.instruments = NULL;
+    sf.drums = NULL;
+    sf.sfx = NULL;
+    LL_FOREACH(xmlNodePtr, node, root->children) {
+        const char *name = XMLSTR_TO_STR(node->name);
+
+        if (strequ(name, "Instruments"))
+            read_instrs_info(&sf, node);
+        if (strequ(name, "Drums"))
+            read_drums_info(&sf, node);
+        if (strequ(name, "Effects"))
+            read_sfx_info(&sf, node);
+    }
+
+    // read match padding if it exists
+    sf.match_padding = NULL;
+    LL_FOREACH(xmlNodePtr, node, root->children) {
+        const char *name = XMLSTR_TO_STR(node->name);
+
+        if (strequ(name, "MatchPadding"))
+            read_match_padding(&sf, node);
+    }
+
+    // emit C source
+
+    FILE *out_c = fopen(filename_out_c, "w");
+    fprintf(out_c, "#include \"soundfont_file.h\"\n\n");
+
+    size_t size = 0;
+    size += emit_c_header(out_c, &sf);
+    size += emit_c_samples(out_c, &sf);
+    size += emit_c_envelopes(out_c, &sf);
+    size += emit_c_instruments(out_c, &sf);
+    size += emit_c_drums(out_c, &sf);
+    size += emit_c_effects(out_c, &sf);
+    emit_c_match_padding(out_c, &sf, size);
+
+    fclose(out_c);
+
+    // emit C header
+
+    FILE *out_h = fopen(filename_out_h, "w");
+    fprintf(out_h,
+            // clang-format off
+           "#ifndef SOUNDFONT_%d_H_"            "\n"
+           "#define SOUNDFONT_%d_H_"            "\n"
+                                                "\n",
+            // clang-format on
+            sf.info.index, sf.info.index);
+
+    fprintf(out_h,
+            // clang-format off
+           "#ifdef _LANGUAGE_ASEQ"              "\n"
+           ".pushsection .fonts, \"\", @note"   "\n"
+           "    .byte %d /*sf id*/"             "\n"
+           ".popsection"                        "\n"
+           "#endif"                             "\n"
+                                                "\n",
+            // clang-format on
+            sf.info.index);
+
+    fprintf(out_h,
+            // clang-format off
+           "#define %s_ID %d"                   "\n"
+                                                "\n"
+           "#define SF%d_NUM_INSTRUMENTS %d"    "\n"
+           "#define SF%d_NUM_DRUMS       %d"    "\n"
+           "#define SF%d_NUM_SFX         %d"    "\n"
+                                                "\n",
+            // clang-format on
+            sf.info.name, sf.info.index, sf.info.index, sf.info.num_instruments, sf.info.index, sf.info.num_drums,
+            sf.info.index, sf.info.num_effects);
+
+    emit_h_instruments(out_h, &sf);
+    emit_h_drums(out_h, &sf);
+    emit_h_effects(out_h, &sf);
```

Comment:
what's the reason these are emitted in the .h file? I only see it included in soundfont_table.h but they don't seem to be used there. I'm not sure I see the point of this header at all actually, shouldn't atblgen have this info and be able to insert the right values directly?

Reply:
These are emitted so they can be included in sequence files, that have yet to be introduced.

I originally had atblgen not care about the headers, but it was a lot of extra code to calculate the quantities that I was already emitting in these headers so I figured it would be better to reuse the work that's already been done.